extending.md

```diff
@@ -0,0 +1,69 @@
+# Extending `smart_open`
+
+This document targets potential contributors to `smart_open`.
+Currently, there are two main directions for extending existing `smart_open` functionality:
+
+1. Add a new transport mechanism
+2. Add a new compression format
+
+## New transport mechanisms
+
+Each transport mechanism lives in its own submodule.
+For example, currently we have:
+
+- `smart_open.file`
+- `smart_open.s3`
+- `smart_open.ssh`
+- ... and others
+
+So, to implement a new transport mechanism, you need to create a new module.
+Your module should expose the following:
+
+```python
+XXX_SCHEMA = ...
+"""The name of the mechanism, e.g. s3, ssh, etc.
+
+This is the part that goes before the `://` in a URL, e.g. `s3://`."""
+
+def parse_uri(uri_as_str):
+    """Parse the specified URI into a dict.
+
+    At a bare minimum, the dict must have `schema` member.
+    """
+    return dict(schema=XXX_SCHEMA, ...)
+
+
+def open_uri(uri_as_str, mode, transport_params):
+    """Return a file-like object pointing to the URI."""
+    ...
+```
+
+Have a look at the existing mechanisms to see how they work.
+You may define other functions and classes as necessary for your implementation.
+
+Once your module is working, register it in the `smart_open/smart_open_lib.py` file.
```

Comment:
I think it would be helpful to document what other documentation needs updating when adding a new transport mechanism. The instances I am aware of are README examples, the setup.py description, and smart_open.open.__doc__.